src/main/java/bio/terra/service/job/StairwayExceptionFieldsFactory.java

```diff
@@ -0,0 +1,103 @@
+      fieldExceptionClassName = Optional.of(ex.getClass().getName());
+    }
+
+    if (stepName.isEmpty()) {
+      return new StairwayExceptionFields()
+          .setClassName(fieldExceptionClassName.orElse(JobExecutionException.class.getName()))
+          .setDataRepoException(true)
+          .setMessage(getJobExceptionMessage(ex))
+          .setErrorDetails(getJobExceptionDetails(ex));
+    }
+
+    return new StairwayExceptionFields()
+        .setClassName(fieldExceptionClassName.orElse(StepExecutionException.class.getName()))
+        .setDataRepoException(true)
+        .setMessage(getStepExceptionMessage(stepName.get(), ex))
+        .setErrorDetails(getStepExceptionDetails(ex));
+  }
+
+  private static Optional<String> getStepNameFromStairwayException(Exception ex) {
+    String failedStepName = null;
+    for (StackTraceElement ste : ex.getStackTrace()) {
+      try {
+        Class<?> cls = Class.forName(ste.getClassName());
+        if (Step.class.isAssignableFrom(cls)) {
+          // Java 17 will make this casting unnecessary
+          failedStepName = cls.getSimpleName();
+          break;
+        }
+      } catch (ClassNotFoundException e) {
+        throw new ExceptionSerializerException(
+            "All steps in a stacktrace should be able to be found", e);
+      }
+    }
+    return Optional.ofNullable(failedStepName);
+  }
+
+  private static String getJobExceptionMessage(Exception exception) {
+    if (StringUtils.isNotBlank(exception.getMessage())) {
+      return exception.getMessage();
+    }
+    return String.format(
+        "Encountered a %s while running the job", exception.getClass().getSimpleName());
+  }
+
+  private static List<String> getJobExceptionDetails(Exception exception) {
+    if (ErrorReportException.class.isAssignableFrom(exception.getClass())) {
+      ErrorReportException errorReportException = (ErrorReportException) exception;
+      return errorReportException.getCauses();
+    } else {
+      List<String> details = new ArrayList<>();
+      details.add("The job failed, but not while running a step");
+      details.add(CONTACT_TEAM_MESSAGE);
+      return details;
+    }
+  }
+
+  private static String getStepExceptionMessage(String stepName, Exception exception) {
+    if (StringUtils.isNotBlank(exception.getMessage())) {
+      return exception.getMessage();
+    }
+    return String.format(
+        "Encountered %s while running %s.", exception.getClass().getSimpleName(), stepName);
+  }
+
+  private static List<String> getStepExceptionDetails(Exception exception) {
+    if (ErrorReportException.class.isAssignableFrom(exception.getClass())) {
+      ErrorReportException errorReportException = (ErrorReportException) exception;
+      return errorReportException.getCauses();
+    } else {
+      List<String> details = new ArrayList<>();
+      if ("IamUnauthorizedException".equals(exception.getClass().getSimpleName())) {
```

Comment:
IMO using `String.equals()` like this makes the code harder to read and doesn't provide a benefit, as `exception.getClass()` is never `null`.
```suggestion
      if (exception.getClass().getSimpleName().equals("IamUnauthorizedException")) {
```